### PR TITLE
✨ [experimental] Support for new `--format probe`

### PR DESCRIPTION
**Pull request by laurentsimon**
Based upon https://github.com/ossf/scorecard/pull/2944 which needs to be merged first.

```release-note
Support for new `--format probe`
```

### CODE REVIEW

File: pkg/scorecard.go
Comment (spencerschrock) [inline pkg/scorecard.go:170]: If the probes have an error this will also affect non-probe users.

File: pkg/scorecard_result.go
Comment (spencerschrock) [inline pkg/scorecard_result.go:54]: Should we stick with `Findings` for consistency?

**Review by spencerschrock — CHANGES_REQUESTED** [review]

File: pkg/json_probe_results.go
Comment (laurentsimon) [reply pkg/json_probe_results.go:52]: I did not just because 
- I wanted some feedback o nest vs flat structure (one of your earlier comment). 
- I don't know if this is needed for practical purposes, so I did not want to over-commit to it right away

I added it back. It will show as:
```json
"metadata": {
    "repository.defaultBranch": "main",
    "repository.host": "github.com",
    "repository.name": "ossf/scorecard-action",
    "repository.sha1": "7fe8d54f4acf26efc826129c396173562e2c766c",
    "repository.uri": "github.com/ossf/scorecard-action"
  }
```

Let me know if you'd like me to keep it or not. If not, I'll create a PR for tracking so that we can re-visit

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:103]: +1, I will not export it for now. I think I'll need this function to run scorecard in "probe mode" vs "check mode". For now I'll make it private.

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:127]: Good catch. I will return an error only if the retuned error is not `clients.ErrUnsupportedFeature`. If the errors is `clients.ErrUnsupportedFeature`, I'll set the defaultBranch to "unknown". It's no ideal because the fix will be incorrect. This has some impact on Scorecard run on PR (which we don't official support). Created https://github.com/ossf/scorecard/issues/3069 for tracking.

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:170]: Good point. I think we can discard the error. The errors will all be part of the findings so we don't need the returned err, see https://github.com/ossf/scorecard/blob/main/probes/zrunner/runner.go#L34-L45

Ill add a comment to explain why we discard the err.

Let me know if this is enough.

File: pkg/scorecard_result.go
Comment (laurentsimon) [reply pkg/scorecard_result.go:54]: Done.

File: checker/raw_result.go
Comment (spencerschrock) [reply checker/raw_result.go:46]: I just meant sticking it in a struct like all the other fields in `RawResults`. So we can add more to it later in a non-breaking way if we had to. Do you ever envision any metadata that would be non-string?

```
// If you have a better name use it haha.
type MetadataData {
	Metadata map[string]string
}
```

Comment (spencerschrock) [reply checker/raw_result.go:46]: Although stepping back, does this metadata really fit with RawResults?

There's already `Metadata []string` in the `ScorecardResult`struct, which is a bit different since it's user supplied.
```
type ScorecardResult struct {
	Repo       RepoInfo
	Date       time.Time
	Scorecard  ScorecardInfo
	Checks     []checker.CheckResult
	RawResults checker.RawResults
	Metadata   []string
}
```

File: pkg/json_probe_results.go
Comment (spencerschrock) [reply pkg/json_probe_results.go]: If you don't want to over-commit right now I would remove the `Metadata` field from `JSONScorecardProbeResult`

File: pkg/scorecard.go
Comment (spencerschrock) [reply pkg/scorecard.go]: Is this just in preparation for future argument changes? Does this need to be added this PR?

File: pkg/scorecard.go
Comment (spencerschrock) [inline pkg/scorecard.go:176]: linter is failing you here for ignoring err, which you should be checking. 

My original feedback meant more along the lines of dont run the probes if you're not on experimental

File: checker/raw_result.go
Comment (spencerschrock) [reply checker/raw_result.go:46]: Sorry to nitpick, just trying to plan in a way that avoids potential breaking changes in the future to the top level struct

**Review by spencerschrock — COMMENTED** [review]
Once we answer some of these questions, it would be good to get some tests written

File: checker/raw_result.go
Comment (laurentsimon) [reply checker/raw_result.go:46]: no that's fine. I could make it a `Metadata map[string]any`? I would like to keep all values strings in practice, but defining it as a map to `any` would give us flexibility if we want to. Wdut?

File: pkg/json_probe_results.go
Comment (laurentsimon) [reply pkg/json_probe_results.go]: ok, rolling back then.

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go]: does not have to, but it's private now.

File: checker/raw_result.go
Comment (spencerschrock) [reply checker/raw_result.go:46]: I still think putting it in a struct gives us the most flexibility to change things later:
```go
// If you have a better name use it haha.
type MetadataData {
	Metadata map[string]string
}
```

File: pkg/json_probe_results.go
Comment (spencerschrock) [inline pkg/json_probe_results.go:34]: if we're not populating the `Metadata` field yet, let's drop it from the struct. Can always add it later.

File: pkg/scorecard.go
Comment (spencerschrock) [inline pkg/scorecard.go:130]: can you wrap instead of ignoring?

File: pkg/scorecard.go
Comment (spencerschrock) [inline pkg/scorecard.go:124]: my understanding was the wrapcheck only complained about external errors?

File: pkg/scorecard.go
Comment (spencerschrock) [inline pkg/scorecard.go:105]: this still seems like an unneeded change to me, but not going to block on it.

File: pkg/json_probe_results.go
Comment (spencerschrock) [inline pkg/json_probe_results.go:38]: please add tests

File: pkg/json_probe_results.go
Comment (laurentsimon) [reply pkg/json_probe_results.go:34]: done, thanks.

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:124]: done.

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:130]: done

File: pkg/scorecard.go
Comment (laurentsimon) [reply pkg/scorecard.go:105]: removed.

File: pkg/json_probe_results.go
Comment (laurentsimon) [reply pkg/json_probe_results.go:38]: created https://github.com/ossf/scorecard/issues/3110 for tracking. Can you comment on the issue what kind of test would be useful? (This PJSON function just prints to an io)

**Review by spencerschrock — APPROVED** [review]
Thanks. There's a linter warning about lll

**Pull request merged**
